File: Tools/bin2h/bin2h.py
#!/usr/bin/env python3
import sys

def main():
    if len(sys.argv) > 3 or len(sys.argv) < 2:
        print(f"Usage: bin2h.py\n")
        print(f"\t [Output] - accepts stdin as input file\n")
        print(f"\t [Input] [Output] - accepts input file\n")
        sys.exit(1)

    out_path = ""
    if len(sys.argv) == 3:
        in_path = sys.argv[1]
        out_path = sys.argv[2]
        with open(in_path, "rb") as f:
            data = f.read()
    else:
        out_path = sys.argv[1]
        data = sys.stdin.buffer.read()

    var_name = out_path.replace(".", "_").replace("/", "_")

    lines = []
    lines.append(f"static const unsigned char {var_name}[{len(data)}] = {{\n")

    for i, b in enumerate(data):
        lines.append(f"0x{b:02x},")
        if (i + 1) % 12 == 0:
            lines.append("\n")

    lines.append("\n};")
    
    with open(out_path, "w") as out:
        out.write("".join(lines))

if __name__ == "__main__":
    main()

File: Tools/bin2h/bin2h.c
#include <stdio.h>
#include <stdlib.h>
#ifdef _WIN32
    #include <io.h>
    #include <fcntl.h>
#endif

void PrintUsage()
{
    fprintf(stderr, "Usage: \n");
    fprintf(stderr, "\tbin2h [Input] [Output]\n");
    fprintf(stderr, "\tbin2h [Output] - reads from stdin\n");
    fprintf(stderr, "using '-' for output will forward output into stdout");
}

int main(int argc, char* argv[]) {
    #ifdef _WIN32
    _setmode(_fileno(stdin), _O_BINARY);
    _setmode(_fileno(stdout), _O_BINARY);
    #endif

    FILE* in = stdin;
    FILE* out = stdout;
    short closeIn = 0;

    if(argc == 2) {
        if(argv[1][0] == '-') {
            if(argv[1][1] == 'h') {
                PrintUsage();
                return 0;
            } else {
                if (isatty(fileno(in))) {
                    fprintf(stderr, "Error: Can't read from terminal.");
                    return 1;
                }
            }
        } else {
            if (isatty(fileno(in))) {
                fprintf(stderr, "Error: Can't read from terminal.");
                return 1;
            }
            out = fopen(argv[1], "wb");
        }
    } else if(argc == 3) {
        in = fopen(argv[1], "rb");
        if (!in) {
            fprintf(stderr, "bin2h: Invalid input file name");
            return 1;
        }

        // '-' specifies a pipe
        if(argv[2][0] != '-') {
            out = fopen(argv[2], "wb");
            if(!out) {
                fprintf(stderr, "bin2h: Unable to open output file");
                return 1;
            }
        }
        closeIn = 255;
    } else {
        PrintUsage();
        return 0;
    }

    // Write as C array
    int c, i = 0;

    fprintf(out, "static const unsigned char data[] = {\n");
    while ((c = fgetc(in)) != EOF) {
        fprintf(out, "0x%02x, ", c);
        if ((i + 1) % 12 == 0) fprintf(out, "\n");

        ++i;
    }
    fprintf(out, "\n};");

    // Close if it's not an std file stream
    if (closeIn) fclose(in);

    return 0;
}